README.markdown
CouponCode
==========

A 'CouponCode' is a type of code that will be passed *in printed form* to
someone who will be expected to type it into a web page or other application.
The codes are random strings which are designed to be easy for the recipient to
type accurately into a web form.

The following features make the codes well suited to manual transcription:

* The codes are not case sensitive.
* Not all letters and numbers are used, so if a person enters the letter 'O' we
  can automatically correct it to the digit '0' (similarly for I => 1, S => 5,
  Z => 2).
* The 4th character of each part is a checkdigit, so client-side scripting can
  be used to highlight parts which have been mis-typed, before the code is even
  submitted to the application's back-end validation.
* The checkdigit algorithm takes into account the position of the part being
  keyed. So for example '1K7Q' might be valid in the first part but not in the
  second so if a user typed the parts in the wrong boxes then their error could
  be highlighted.
* The code generation algorithm avoids 'undesirable' codes. For example any
  code in which transposed characters happen to result in a valid checkdigit
  will be skipped. Any generated part which happens to spell an 'inappropriate'
  4-letter word (e.g.: 'P00P') will also be skipped.

Algorithm::CouponCode
---------------------

This git repository contains:

* the Perl [CPAN module](http://search.cpan.org/dist/Algorithm-CouponCode/)
which implements the code generation algorithm and the server-side validation
* a [Javascript implementation](http://search.cpan.org/dist/Algorithm-CouponCode/html/index.html)
of the validator function, in the form of a jQuery plugin, which can be
included in your web application to do client-side validation and highlighting
of errors

Alternative implementations
---------------------------

The code generation and server-side validation routines have been ported to
other languages (_Note: a listing here does not constitute an endorsement nor a
guarantee of compatibility_):


* [Javascript for Node.js](https://www.npmjs.com/package/coupon-code)
* [Ruby](https://rubygems.org/gems/coupon_code/versions/0.0.1)
* [PHP](https://github.com/atelierdisko/coupon_code)


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

1.005     2015-11-25 21:09:15+13:00 Pacific/Auckland
  - update docs to link to ports to other languages
  - allow JS plugin to pre-populate fields from value attribute

1.004     2013-08-02 10:07:30 Pacific/Auckland
  - Change dependency from Digest::SHA1 to Digest::SHA

1.003     2011-05-23 21:27:20 Pacific/Auckland
  - Javascript change: allow paste of full code into first field
  - more Javascript fixes for IE compatibility

1.002     2011-05-22 20:19:15 Pacific/Auckland
  - no code changes - just CSS tweaks

1.001     2011-05-20 00:07:17 Pacific/Auckland
  - POD update to clarify intent of validation
  - fix Javascript to work on older versions of IE
  - add Javascript option to set intial focus
  - add CSS + icons to improve accessibility

1.000     2011-04-13 12:26:35 Pacific/Auckland
  - Remove redundant reference to Try::Tiny

0.900     2011-04-12 13:55:43 Pacific/Auckland
  - Initial release to CPAN


base32-coding.pl
#!/usr/bin/perl
#
# http://www.crockford.com/wrmg/base32.html
#

use strict;
use warnings;

my %algorithm = (
    a  => \&algorithm_a,                            # the original
    b  => \&algorithm_b,                            # mod 31 instead of 29
    c  => \&algorithm_c,                            # use 5 LSB rather than mod
    d  => \&algorithm_d,                            # rotation + XOR
);
my %symbol_set = (
    A => '0123456789ABCDEFGHJKMNPQRSTVWXYZ',
    B => '0123456789ABCDEFGHJKMNPQRSTVWXZY',       # X & Y swapped
    C => '0123456789ABCDEFGHJKLMNPQRSTVWXY',       # Add L, Drop Z
    D => '0123456789ABCDEFGHJKLMNPQRTUVWXY',       # Add L & U, Drop S & Z
);
my($sym, @sym, $check_sub, %tally, %sully);
my $cuss_words = make_cuss_regex();

foreach my $alg ( sort keys %algorithm) {
    foreach my $set ( sort keys %symbol_set) {

        print "Testing Algorithm $alg with symbol set $set\n";
        $check_sub = $algorithm{$alg};
        $sym       = $symbol_set{$set};
        @sym       = split //, $sym;

        my $total     = 0;
        my $bad_swaps = 0;
        my $bad_reads = 0;
        my $bad_pos   = 0;
        my $cuss      = 0;
        foreach my $x ( @sym ) {
            foreach my $y ( @sym ) {
                foreach my $z ( @sym ) {
                    my $code = "$x$y$z";
                    foreach my $pos (0, 1, 2) {
                        my $pcode = $code . $check_sub->($code, $pos);
                        $total++;
                        $bad_swaps += check_swaps($pcode, $pos);
                        $bad_reads += check_reads($pcode, $pos);
                        $bad_pos   += check_pos($pcode, $pos);
                        $cuss      += check_cuss($pcode, $pos);
                    }
                }
            }
        }

        printf("Total codes generated: %5u\n", $total);
        printf(
            "Total bad swaps:       %5u ( %5.2f%%)\n",
            $bad_swaps, 100 * $bad_swaps / $total
        );
        printf(
            "Total bad reads:       %5u ( %5.2f%%)\n",
            $bad_reads, 100 * $bad_reads / $total
        );
        printf(
            "Total bad pos:         %5u ( %5.2f%%)\n",
            $bad_pos, 100 * $bad_pos / $total
        );
        my $rate = sprintf(
            "%4.2f", 100 * ($bad_swaps + $bad_reads + $bad_pos) / $total
        );
        printf("False positive rate:   %s%%\n\n", $rate);
        printf("Total cuss words:      %5u\n", $cuss);

        $tally{$alg}{$set} = $rate;
        $sully{$alg}{$set} = $cuss;
    }
}

# Print accumulated results

print "              ";
print "    Set $_     " foreach ( sort keys %symbol_set );
print "\n";

foreach my $alg ( sort keys %algorithm ) {
    printf("algorithm_%-2s", $alg);
    foreach my $set ( sort keys %symbol_set ) {
        printf("%8s%% (%2u)", $tally{$alg}{$set}, $sully{$alg}{$set});
    }
    print "\n";
}

exit;


sub check_swaps {
    my($orig, $pos) = @_;

    my $bad = 0;
    my($a, $b, $c, $d) = split //, $orig;
    foreach my $code (
        "$b$a$c$d",
        "$a$c$b$d",
        "$a$b$d$c",
    ) {
        next if $code eq $orig;
        if($check_sub->(substr($code, 0, 3), $pos) eq substr($code, 3, 1)) {
            $bad++;
        }
    }
    return $bad;
}


sub check_reads {
    my($orig, $pos) = @_;

    my @try = map { [ $_ ] } split //, $orig;
    foreach (@try) {
        push @$_, '5' if $_->[0] eq 'S';
        push @$_, 'S' if $_->[0] eq '5';
        push @$_, '2' if $_->[0] eq 'Z';
        push @$_, 'Z' if $_->[0] eq '2';
    }
    return 0 unless grep { @$_ > 1 } @try;
    my $bad = 0;
    foreach my $a ( @{ $try[0] } ) {
        foreach my $b ( @{ $try[1] } ) {
            foreach my $c ( @{ $try[2] } ) {
                foreach my $d ( @{ $try[3] } ) {
                    my $code = "$a$b$c$d";
                    next if $code eq $orig;
                    if($check_sub->(substr($code, 0, 3), $pos) eq substr($code, 3, 1)) {
                        $bad++;
                    }
                }
            }
        }
    }
    return $bad;
}


sub check_pos {
    my($orig, $pos) = @_;

    my $bad  = 0;
    my $code = substr($orig, 0, 3);
    foreach my $i ( 0, 1, 2 ) {
        next if $pos == $i;
        if($check_sub->($code, $i) eq substr($orig, 3, 1)) {
            $bad++;
        }
    }
    return $bad;
}


sub check_cuss {
    my($orig, $pos) = @_;

    return 0 unless $orig =~ $cuss_words;
    print "Cuss word: $orig\n";
    return 1;
}


sub make_cuss_regex {
    my $words = join '|', map {
        s/[I1]/[I1]/g;
        s/[O0]/[O0]/g;
        s/[S5]/[S5]/g;
        s/[Z2]/[Z2]/g;
        s/[E3]/[E3]/g;
        $_;
    } map { uc($_) } qw (
        fuck cunt wank wang piss cock shit twat tits fart hell muff dick knob
        arse shag toss slut turd slag crap poop butt feck boob jism jizz
    );

    return qr{\A(?:$words)\z};
}


sub algorithm_a {
    my($data, $pos) = @_;
    my @char = split //, $data;

    my $check = $pos;
    foreach my $i (0..2) {
        my $k = index($sym, $char[$i]);
        $check = $check * 19 + $k;
    }
    return $sym[ $check % 29 ];
}


sub algorithm_b {
    my($data, $pos) = @_;
    my @char = split //, $data;

    my $check = $pos;
    foreach my $i (0..2) {
        my $k = index($sym, $char[$i]);
        $check = $check * 19 + $k;
    }
    return $sym[ $check % 31 ];
}


sub algorithm_c {
    my($data, $pos) = @_;
    my @char = split //, $data;

    my $check = $pos;
    foreach my $i (0..2) {
        my $k = index($sym, $char[$i]);
        $check = $check * 19 + $k;
    }
    return $sym[ $check & 31 ];
}


sub algorithm_d {
    my($data, $pos) = @_;
    my @char = split //, $data;

    my $check = $pos;
    foreach my $i (0..2) {
        my $k = index($sym, $char[$i]);
        $check = ($check + $k) << 3;
        $check = $check ^ ($check >> 4);
    }
    return $sym[ $check & 31 ];
}




html/jquery.couponcode.js
/*
 * couponCode - jQuery plugin
 *
 * Copyright (c) 2011 Grant McLean (grant@mclean.net.nz)
 *
 * Dual licensed under the MIT and GPL licenses:
 *   http://www.opensource.org/licenses/mit-license.php
 *   http://www.gnu.org/licenses/gpl.html
 *
 */

(function($) {

var SYMBOL_SET = '0123456789ABCDEFGHJKLMNPQRTUVWXY';
var BAD_SYMBOL = new RegExp('[^' + SYMBOL_SET + ']');

$.fn.couponCode = function(options) {
    return this.each(function() {
        $.fn.couponCode.build(this, options);
    });
};

$.fn.couponCode.build = function(base_entry, options) {
    var self    = $.extend({}, $.fn.couponCode.defaults, options);
    self.focus  = null;
    self.inputs = [];
    self.flags  = [];
    self.parts  = parseInt(self.parts, 10);
    if(isNaN(self.parts) || self.parts < 1 || self.parts > 6) {
        alert("CouponCode 'parts' must be in range 1-6");
        return;
    }

    var start_val = $(base_entry).val();
    var name    = base_entry.name;
    var id      = base_entry.id;
    var wrapper = $( $(base_entry).wrap('<span class="jq-couponcode" />').parent()[0] );
    wrapper[0].removeChild(base_entry);
    var hidden  = $('<input type="hidden">').attr({ 'name': name, 'id': id });
    wrapper.append(hidden);
    var inner   = $('<span class="jq-couponcode-inner" />');
    for(var i = 0; i < self.parts; i++) {
        if(i > 0) {
            inner.append($('<span class="jq-couponcode-sep" />').text(self.separator));
        }
        self.inputs[i] = $('<input type="text" class="jq-couponcode-part" />');
        inner.append(self.inputs[i]);
    }
    $( self.inputs ).each(function(i, input) {
        input
        .keydown(function() { setTimeout(function() { validate(i); }, 5 ); } )
        .blur(function() { self.focus = null; validate(i); } )
        .focus( function() { self.focus = i; } );
    });
    self.inputs[0].on('paste', function() {
        setTimeout(function() { set_parts(self.inputs[0].val()); }, 2);
    });

    if(start_val.length > 0) {
        set_parts(start_val);
    }
    wrapper.append(inner);
    if(self.setFocus) {
        self.inputs[0].focus();
    }

    function set_parts(code) {
        var parts = code.split('-');
        for(var i = 0; i < self.parts; i++) {
            self.inputs[i].val( parts[i] || '');
            validate(i);
        }
    }

    function validate(index) {
        var input  = self.inputs[index];
        var result = validate_one_field(input, index);
        if(result === true) {
            input.removeClass('jq-couponcode-bad');
            input.addClass('jq-couponcode-good');
            self.flags[index] = 1;
        }
        else {
            input.removeClass('jq-couponcode-good');
            input.removeClass('jq-couponcode-bad');
            if(result === false) {
                input.addClass('jq-couponcode-bad');
            }
            self.flags[index] = 0;
        }
        update_hidden_field();
    }

    function validate_one_field(input, index) {
        var val = input.val();
        var focussed = (self.focus === index);
        if(val == '') { return; }
        var code = clean_up( val );
        if(code.length > 4 || BAD_SYMBOL.test(code)) {
            return false;
        }
        if(code.length < 4) {
            return focussed ? null : false;
        }
        if(code.charAt(3) != checkdigit(code, index + 1)) {
            return false;
        }
        if(val != code) {
            input.val(code);
        }
        if(focussed && (index < self.parts - 1)) {
            self.inputs[index + 1].focus();
            self.focus = index + 1;
        }
        return true;
    }

    function clean_up(code) {
        code = code.toUpperCase()
               .replace(/ /g, '')
               .replace(/O/g, '0')
               .replace(/I/g, '1')
               .replace(/S/g, '5')
               .replace(/Z/g, '2');
        return code;
    }

    function checkdigit(data, pos) {
        var check = pos;
        for(var i = 0; i < 3; i++) {
            var k = SYMBOL_SET.indexOf( data.charAt(i) );
            check = check * 19 + k;
        }
        return SYMBOL_SET.charAt(check % 31);
    }

    function update_hidden_field() {
        var parts    = [];
        var old_code = hidden.val();
        $( self.inputs ).each(function(i, input) {
            parts.push(input.val());
        });
        var new_code = parts.join('-');
        if(old_code == new_code) {
            return;
        }
        hidden.val(new_code);
        if(self.onChange) {
            var good_parts = 0;
            $( self.flags ).each(function(i, val) { good_parts = good_parts + val; });
            self.onChange(good_parts === self.parts);
        }
    }

};

$.fn.couponCode.defaults = {
    parts     : 3,
    separator : '-',
    setFocus  : false
};

})(jQuery);


lib/Algorithm/CouponCode.pm
package Algorithm::CouponCode;

=head1 NAME

Algorithm::CouponCode - Generate and validate 'CouponCode' strings

=cut

use 5.010;
use warnings;
use strict;


use Exporter     qw(import);
use Digest::SHA  qw(sha1);


our @EXPORT_OK   = qw(cc_generate cc_validate make_bad_regex);

my $sym_str      = '0123456789ABCDEFGHJKLMNPQRTUVWXY';
my @sym          = split //, $sym_str;
my $urandom_path = '/dev/urandom';
my $have_urandom = -r $urandom_path;
my $bad_regex    = make_bad_regex();


sub cc_generate {
    my %arg = @_;

    my $parts     = $arg{parts} || 3;
    my $plaintext = $arg{plaintext} // _random_plaintext();
    my $bad_words = $arg{bad_regex} || $bad_regex;
    my($sha1_hash, @code);

    RANDOM_HASH: {
        $sha1_hash = sha1($plaintext);
        my @bytes  = map { ord($_) & 31 } split //, $sha1_hash;
        TRY_PART: while(@code < $parts) {
            if(@bytes < 3) {
                $plaintext = $sha1_hash;
                redo RANDOM_HASH;
            }
            my $i    = @code + 1;
            my $str  = join '', map { $sym[shift @bytes] } (0, 1, 2);
            my $part = $str . _checkdigit_alg_1($str, $i);
            next TRY_PART if $part =~ $bad_words;
            next TRY_PART if _valid_when_swapped($part, $i);
            push @code, $part;
        }
    }

    return join '-', @code;
}


sub cc_validate {
    my %arg = @_;

    my $code   = $arg{code} or return;
    my $parts  = $arg{parts} // 3;

    $code =  uc($code);
    $code =~ s{[^0-9A-Z]+}{}g;
    $code =~ tr{OIZS}{0125};
    my(@parts) = $code =~ m{([0-9A-Z]{4})}g;
    return unless scalar(@parts) == $parts;

    foreach my $i (1..$parts) {
        my($str, $check) = $parts[$i - 1] =~ m{^(...)(.)};
        return unless $check eq _checkdigit_alg_1($str, $i);
    }
    return join '-', @parts;
}


sub make_bad_regex {
    my %arg = @_;

    my @word_list = _default_bad_word_list();
    if($arg{words}) {
        if($arg{mode} && $arg{mode} eq 'replace') {
            @word_list = @{ $arg{words} };
        }
        else {
            push @word_list, @{ $arg{words} };
        }
    }
    my $words = join '|', map {
        $_ = uc($_);
        s/[I1]/[I1]/g;
        s/[O0]/[O0]/g;
        s/[S5]/[S5]/g;
        s/[Z2]/[Z2]/g;
        s/[E3]/[E3]/g;
        s/[A4]/[A4]/g;
        $_;
    } @word_list;
    return qr{\A(?:$words)\z};
}


sub _default_bad_word_list {
    # Yay for rot13
    return map { my $w = $_; $w =~ tr/A-Z/N-ZA-M/; $w } qw{
        SHPX PHAG JNAX JNAT CVFF PBPX FUVG GJNG GVGF SNEG URYY ZHSS QVPX XABO
        NEFR FUNT GBFF FYHG GHEQ FYNT PENC CBBC OHGG SRPX OBBO WVFZ WVMM CUNG
    };
}


sub _random_plaintext {
    if($have_urandom) {
        open my $fh, '<', $urandom_path or die "open($urandom_path): $!";
        sysread $fh, my $buf, 8;
        return $buf;
    }
    else {
        return $$ . localtime() . rand();
    }
}


sub _checkdigit_alg_1 {
    my($data, $pos) = @_;
    my @char = split //, $data;

    my $check = $pos;
    foreach my $i (0..2) {
        my $k = index($sym_str, $char[$i]);
        $check = $check * 19 + $k;
    }
    return $sym[ $check % 31 ];
}


sub _valid_when_swapped {
    my($orig, $pos) = @_;

    my($a, $b, $c, $d) = split //, $orig;
    foreach my $code (
        "$b$a$c$d",
        "$a$c$b$d",
        "$a$b$d$c",
    ) {
        next if $code eq $orig;
        if(_checkdigit_alg_1(substr($code, 0, 3), $pos) eq substr($code, 3, 1)) {
            return 1;
        }
    }
    return 0;
}

1;
__END__

=pod

=head1 SYNOPSIS

  use Algorithm::CouponCode qw(cc_generate cc_validate);

  print cc_generate(parts => 3);  # generate a 3-part code

  my $valid_code = cc_validate(code => $code, parts => 3) or die "Invalid code";

=head1 DESCRIPTION

A 'Coupon Code' is made up of letters and numbers grouped into 4 character
'parts'.  For example, a 3-part code might look like this:

  1K7Q-CTFM-LMTC

Coupon Codes are random codes which are easy for the recipient to type
accurately into a web form.  An example application might be to print a code on
a letter to a customer who would then enter the code as part of the
registration process for web access to their account.

Features of the codes that make them well suited to manual transcription:

=over 4

=item *

The codes are not case sensitive.

=item *

Not all letters and numbers are used, so if a person enters the letter 'O' we
can automatically correct it to the digit '0' (similarly for I => 1, S => 5, Z
=> 2).

=item *

The 4th character of each part is a checkdigit, so client-side scripting can
be used to highlight parts which have been mis-typed, before the code is even
submitted to the application's back-end validation.

=item *

The checkdigit algorithm takes into account the position of the part being
keyed.  So for example '1K7Q' might be valid in the first part but not in the
second so if a user typed the parts in the wrong boxes then their error could
be highlighted.

=item *

The code generation algorithm avoids 'undesirable' codes. For example any code
in which transposed characters happen to result in a valid checkdigit will be
skipped.  Any generated part which happens to spell an 'inappropriate' 4-letter
word (e.g.: 'P00P') will also be skipped.

=back

The Algorithm-CouponCode distribution includes a Javascript implementation of
the validator function, in the form of a jQuery plugin.  You can include this
in your web application to do client-side validation and highlighting of
errors.

I<Note> the cc_validate function and the Javascript plugin only validate that
the code is 'well-formed' (i.e.: each part has a valid checkdigit).  Checking
whether the code is in fact 'valid' is left up to your application and would
typically involve looking for the code in a database.  If you use the
Javascript plugin, you might choose to tweak the CSS to keep the red
highlighting of checksum errors but remove the green highlighting which might
imply the code was correct.

=head2 Randomness and Uniqueness

The code returned by C<cc_generate()> is random, but not necessarily unique.
If your application requires unique codes, it is your responsibility to
avoid duplicates (for example by using a unique index on your database column).

The codes are generated using a SHA1 cryptographic hash of a plaintext.  If you
do not supply a plaintext, one will be generated for you (using /dev/urandom if
available or Perl's C<rand()> function otherwise).  In the event that an
'inappropriate' code is created, the generated hash will be used as a
plaintext input for generating a new hash and the process will be repeated.

Each 4-character part encodes 15 bits of random data, so a 3-part code will
incorporate 45 bits making a total of 2^45 (approximately 35 trillion) unique
codes.


=head1 EXPORTS

The following functions can be exported from the C<Algorithm::CouponCode>
module. No functions are exported by default.

=head2 cc_generate( options )

Returns a coupon code as a string of 4-character parts separated by '-'
characters.  The following optional named parameters may be supplied:

=over 4

=item parts

The number of parts desired.  Must be a number in the range 1 - 6.  Default is
3.

=item plaintext

A byte string which will be hashed using L<Digest::SHA> to produce the code.
If you do not supply your own plaintext then a random one will be generated for
you.

=item bad_regex

You can supply a regular expression for matching 4-letter words which should
not appear in generated output.  The C<make_bad_regex()> helper function can
be used to turn a list of words into a suitable regular expression.

=back

=head2 cc_validate( options )

Takes a code, cleans it up and validates the checkdigits.  Returns the
normalised (and untainted) version of the code on success or undef on error.
The following named parameters may be supplied:

=over 4

=item code

The code to be validated.  The parameter is mandatory.

=item parts

The number of parts you expect the code to contain.  Default is 3.

=back

=head2 make_bad_regex( options )

This function is used to compile a list of 4-letter words into a regular
expression suitable for passing to the C<cc_generate()> function.  You would
only need to do this if you wished to augment or replace the default list of
undesirable words.  The following named parameters may be supplied:

=over 4

=item words

A reference to an array of 4-letter words.  The parameter is mandatory.

=item mode

Either the word 'add' or the word 'replace'.  Default is 'add'.

=back


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Algorithm::CouponCode

You can also look for information at:

=over 4

=item * Github project issue tracker

L<https://github.com/grantm/Algorithm-CouponCode/issues>

=item * Source Code Respository

L<http://github.com/grantm/Algorithm-CouponCode>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Algorithm::CouponCode>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Algorithm::CouponCode>

=item * metacpan

L<https://metacpan.org/release/Algorithm-CouponCode>

=back


=head1 ALTERNATIVE IMPLEMENTATIONS

The code generation and server-side validation routines have been ported to
other languages (I<Note: a listing here does not constitute an endorsement nor
a guarantee of compatibility>):

=over 4

=item *

Javascript for Node.js: L<https://www.npmjs.com/package/coupon-code>

=item *

Ruby: L<https://rubygems.org/gems/coupon_code/versions/0.0.1>

=item *

PHP: L<https://github.com/atelierdisko/coupon_code>

=back


=head1 COPYRIGHT AND LICENSE

Copyright 2011-2015 Grant McLean C<< <grantm@cpan.org> >>

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut
